A panic is triggered when the miner node is restarted

#### System information

Geth version: `v0.3.2`
OS & Version: Windows/Linux/OSX
Commit hash : a80d7c9521822d44d69d8ce9e70d16e24e91d63d

#### Expected behaviour
The miner node can be started normally.

#### Actual behaviour
The miner node crashed when it was restarted.

#### Steps to reproduce the behaviour
1. Start the network of 7 CNs, run `docker_privnet_start_seven`
2. Stop several miner nodes
3. Start miner node

#### Backtrace

````
2025-05-28T09:08:54.251Z	DEBUG	received message	{"type": "ChangeView", "from": 5, "height": 114, "view": 0, "my_height": 114, "my_view": 0}

2025-05-28T09:08:54.251Z	INFO	received ChangeView	{"validator": 5, "reason": "Timeout", "new view": 1}

2025-05-28T09:08:54.251Z	INFO	changing dbft view	{"height": 114, "view": 1, "index": 2, "role": "Backup"}

2025-05-28T09:08:54.251Z	DEBUG	reset timer	{"h": 114, "v": 1, "delay": "40s"}

INFO [05-28|09:08:54.251] Change view detected, waiting for new sealing task to be submitted by miner "old view"=0 "new view"=1

INFO [05-28|09:08:54.251] Fetching latest sealing proposal         "desired number"=114

INFO [05-28|09:08:54.313] Sealing proposal updated                 number=114 sealhash=4b143b..9c8d68 "parent hash"=e97288..2200b1 txs=0

INFO [05-28|09:08:54.313] Start dBFT process for updated sealing work index=114 view=1

2025-05-28T09:08:54.313Z	DEBUG	received message	{"type": "PrepareRequest", "from": 1, "height": 114, "view": 1, "my_height": 114, "my_view": 1}

2025-05-28T09:08:54.313Z	INFO	received PrepareRequest	{"validator": 1, "tx": 0}

panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc602f4]


goroutine 4721 [running]:

github.com/ethereum/go-ethereum/core/state.(*StateDB).Copy(0x0)

	github.com/ethereum/go-ethereum/core/state/statedb.go:691 +0x24

github.com/ethereum/go-ethereum/core/txpool/legacypool.newNoncer(...)

	github.com/ethereum/go-ethereum/core/txpool/legacypool/noncer.go:38

github.com/ethereum/go-ethereum/core/txpool/legacypool.(*LegacyPool).ResetStatic(0x4000ac0c88)

	github.com/ethereum/go-ethereum/core/txpool/legacypool/legacypool.go:380 +0x1a4

github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).verifyPreBlockCb(0x40003b0d88, {0x20fe1e0?, 0x4000244840})

	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:938 +0x4a4

github.com/nspcc-dev/dbft.(*DBFT[...]).createAndCheckBlock(0x2129ac0)

	github.com/nspcc-dev/dbft@v0.3.2/dbft.go:366 +0x58

github.com/nspcc-dev/dbft.(*DBFT[...]).onPrepareRequest(0x2129ac0, {0x210b0b8, 0x40207cefc0})

	github.com/nspcc-dev/dbft@v0.3.2/dbft.go:331 +0x844

github.com/nspcc-dev/dbft.(*DBFT[...]).OnReceive(0x2129ac0, {0x210b0b8, 0x40207cefc0})

	github.com/nspcc-dev/dbft@v0.3.2/dbft.go:269 +0xa30

github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop(0x40003b0d88)

	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:2262 +0x840

created by github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start in goroutine 4376

	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:2093 +0x550

[node.log](https://github.com/user-attachments/files/20480584/node.log)

````

When submitting logs: please submit them as text and not screenshots.

It seems to be related to this #466.

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -1157,6 +1157,9 @@ func (c *DBFT) processPreBlockCb(b dbft.PreBlock[common.Hash]) error {
 		if ctx.IsPrimary() && pre.finalState == nil {
 			receipts = c.sealingReceipts
 		}
+		if len(receipts) != len(pre.transactions) {
+			return fmt.Errorf("number of receipts mismatch: expected %d, actual %d", len(pre.transactions), len(receipts))
+		}
 		// No need to reinitialize static pool since it was already initialized in verifyPreBlockCb.
 		// Reset the static pool after processing all transactions.
 		for i := range pre.transactions {

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -1274,8 +1274,8 @@ func newStaticPool(chain ChainHeaderReader) *legacypool.LegacyPool {
 }
 
 // initStaticPool initializes the static pool with the provided parent header.
-func (c *DBFT) initStaticPool(parent *types.Header) error {
-	return c.staticPool.InitStatic(legacypool.DefaultConfig.PriceLimit, parent, func(addr common.Address, reserve bool) error { return nil })
+func (c *DBFT) initStaticPool(parent *types.Header, state *state.StateDB) error {
+	return c.staticPool.InitStatic(legacypool.DefaultConfig.PriceLimit, parent, state, func(addr common.Address, reserve bool) error { return nil })
 }
 
 // validateDecryptedTx checks the validity of the transaction to determine whether the outer envelope transaction should be replaced.
@@ -1553,7 +1553,7 @@ func (c *DBFT) postBlock(h *types.Header, state *state.StateDB) {
 		c.lastBlockHash = h.Hash()
 		c.lastBlockSealHash, _ = honestSealHash(h) // no error expected, h is verified.
 		c.lastBlockExtra = h.Extra
-		err := c.initStaticPool(h)
+		err := c.initStaticPool(h, state)
 		if err != nil {
 			log.Crit("failed to initialize static pool",
 				"height", c.lastIndex,
@@ -2085,7 +2085,7 @@ func (c *DBFT) Start(chain ChainHeaderWriter) {
 		// miner guarantees that. We can't do it earlier because blocks chain may be out of sync compared to headers
 		// chain, ref. 3721f549.
 		if c.lastIndex == currHeader.Number.Uint64() {
-			err := c.initStaticPool(currHeader)
+			err := c.initStaticPool(currHeader, nil)
 			if err != nil {
 				log.Crit("failed to initialize static pool",
 					"height", c.lastIndex,

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -2125,19 +2125,15 @@ func (c *DBFT) Seal(chain consensus.ChainHeaderReader, b *types.Block, results c
 func (c *DBFT) waitForNewSealingProposal(desiredHeight uint64, updateContext bool) error {
 	log.Info("Fetching latest sealing proposal",
 		"desired number", desiredHeight)
-	var (
-		ok           bool
-		lastProposal *types.Block
-	)
+	var lastProposal *types.Block
 	// Wait here...
 	for {
 		c.lastProposalLock.RLock()
 		if c.lastProposal != nil && c.lastProposal.NumberU64() >= desiredHeight {
 			lastProposal = c.lastProposal
-			ok = true
 		}
 		c.lastProposalLock.RUnlock()
-		if ok {
+		if lastProposal != nil {
 			break
 		}
 		select {

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -1274,8 +1274,8 @@ func newStaticPool(chain ChainHeaderReader) *legacypool.LegacyPool {
 }
 
 // initStaticPool initializes the static pool with the provided parent header.
-func (c *DBFT) initStaticPool(parent *types.Header) {
-	c.staticPool.InitStatic(legacypool.DefaultConfig.PriceLimit, parent, func(addr common.Address, reserve bool) error { return nil })
+func (c *DBFT) initStaticPool(parent *types.Header) error {
+	return c.staticPool.InitStatic(legacypool.DefaultConfig.PriceLimit, parent, func(addr common.Address, reserve bool) error { return nil })
 }
 
 // validateDecryptedTx checks the validity of the transaction to determine whether the outer envelope transaction should be replaced.
@@ -1553,7 +1553,13 @@ func (c *DBFT) postBlock(h *types.Header, state *state.StateDB) {
 		c.lastBlockHash = h.Hash()
 		c.lastBlockSealHash, _ = honestSealHash(h) // no error expected, h is verified.
 		c.lastBlockExtra = h.Extra
-		c.initStaticPool(h)
+		err := c.initStaticPool(h)
+		if err != nil {
+			log.Crit("failed to initialize static pool",
+				"height", c.lastIndex,
+				"hash", c.lastBlockHash,
+				"error", err)
+		}
 
 		// handle DKG
 		if c.lastIndex >= uint64(c.config.dkgEnablingHeight) {
@@ -2079,7 +2085,13 @@ func (c *DBFT) Start(chain ChainHeaderWriter) {
 		// miner guarantees that. We can't do it earlier because blocks chain may be out of sync compared to headers
 		// chain, ref. 3721f549.
 		if c.lastIndex == currHeader.Number.Uint64() {
-			c.initStaticPool(currHeader)
+			err := c.initStaticPool(currHeader)
+			if err != nil {
+				log.Crit("failed to initialize static pool",
+					"height", c.lastIndex,
+					"hash", c.lastBlockHash,
+					"error", err)
+			}
 			if c.lastIndex >= uint64(c.config.dkgEnablingHeight) {
 				c.lock.RLock()
 				ks := c.amevKeystore

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -914,10 +914,6 @@ func (c *DBFT) verifyPreBlockCb(b dbft.PreBlock[common.Hash]) bool {
 	}
 	ethBlock := dbftBlock.ToEthBlock()
 
-	// If is the first view of the block, then initialize static pool with fresh parent.
-	if c.dbft.Context.ViewNumber == 0 {
-		c.initStaticPool(parent)
-	}
 	errs := c.staticPool.Add(dbftBlock.transactions, false, false)
 	for i, err := range errs {
 		if err != nil {
@@ -1557,6 +1553,7 @@ func (c *DBFT) postBlock(h *types.Header, state *state.StateDB) {
 		c.lastBlockHash = h.Hash()
 		c.lastBlockSealHash, _ = honestSealHash(h) // no error expected, h is verified.
 		c.lastBlockExtra = h.Extra
+		c.initStaticPool(h)
 
 		// handle DKG
 		if c.lastIndex >= uint64(c.config.dkgEnablingHeight) {
@@ -2075,6 +2072,14 @@ func (c *DBFT) Start(chain ChainHeaderWriter) {
 				"index", c.lastIndex+1,
 				"err", err.Error())
 			return
+		}
+
+		// Manually initialize static pool in case if waitForNewSealingProposal didn't call
+		// postBlock callback. We're sure that state of the latest block is available by this moment,
+		// miner guarantees that. We can't do it earlier because blocks chain may be out of sync compared to headers
+		// chain, ref. 3721f549.
+		if c.lastIndex == currHeader.Number.Uint64() {
+			c.initStaticPool(currHeader)
 		}
 
 		log.Info("Starting dBFT engine",

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -915,6 +915,7 @@ func (c *DBFT) verifyPreBlockCb(b dbft.PreBlock[common.Hash]) bool {
 	ethBlock := dbftBlock.ToEthBlock()
 
 	errs := c.staticPool.Add(dbftBlock.transactions, false, false)
+	c.staticPool.ResetStatic()
 	for i, err := range errs {
 		if err != nil {
 			log.Warn("proposed PreBlock has invalid transaction",
@@ -925,7 +926,6 @@ func (c *DBFT) verifyPreBlockCb(b dbft.PreBlock[common.Hash]) bool {
 			return false
 		}
 	}
-	c.staticPool.ResetStatic()
 
 	state, receipts, gasUsed, err := c.chain.VerifyBlock(ethBlock, true)
 	if err != nil {

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -2074,12 +2074,24 @@ func (c *DBFT) Start(chain ChainHeaderWriter) {
 			return
 		}
 
-		// Manually initialize static pool in case if waitForNewSealingProposal didn't call
+		// Manually initialize static pool and DKG snapshot in case if waitForNewSealingProposal didn't call
 		// postBlock callback. We're sure that state of the latest block is available by this moment,
 		// miner guarantees that. We can't do it earlier because blocks chain may be out of sync compared to headers
 		// chain, ref. 3721f549.
 		if c.lastIndex == currHeader.Number.Uint64() {
 			c.initStaticPool(currHeader)
+			if c.lastIndex >= uint64(c.config.dkgEnablingHeight) {
+				c.lock.RLock()
+				ks := c.amevKeystore
+				c.lock.RUnlock()
+				err := c.handleDKG(c.dkgSnapshot, ks, currHeader, nil, false)
+				if err != nil {
+					log.Crit("Failed to initialize DKG snapshot",
+						"height", currHeader.Number.Uint64(),
+						"err", err.Error())
+					return
+				}
+			}
 		}
 
 		log.Info("Starting dBFT engine",
@@ -2144,19 +2156,6 @@ func (c *DBFT) waitForNewSealingProposal(desiredHeight uint64, updateContext boo
 			"sealing proposal index", b.NumberU64())
 		ltstHeader := c.chain.GetHeaderByNumber(b.NumberU64() - 1)
 		c.postBlock(ltstHeader, nil)
-	} else if c.lastIndex >= uint64(c.config.dkgEnablingHeight) {
-		// Manually initialize DKG snapshot based on the latest block information
-		// (we're sure that state of the latest block is available by this moment,
-		// miner guarantees that). We can't do it earlier because blocks chain
-		// may be out of sync compared to headers chain, ref. 3721f549.
-		currHeader := c.chain.GetHeaderByNumber(c.lastIndex)
-		c.lock.RLock()
-		ks := c.amevKeystore
-		c.lock.RUnlock()
-		err := c.handleDKG(c.dkgSnapshot, ks, currHeader, nil, false)
-		if err != nil {
-			return fmt.Errorf("failed to initialize DKG snapshot at height %d: %w", currHeader.Number.Uint64(), err)
-		}
 	}
 
 	if b.ParentHash().Cmp(c.lastBlockHash) != 0 {

--- a/core/txpool/legacypool/legacypool.go
+++ b/core/txpool/legacypool/legacypool.go
@@ -325,7 +325,7 @@ func (pool *LegacyPool) Filter(tx *types.Transaction) bool {
 // goroutines will be spun up and the pool deemed operational afterwards.
 func (pool *LegacyPool) Init(gasTip uint64, head *types.Header, reserve txpool.AddressReserver) error {
 	// Do basic initializations about reserver, gas price and state
-	pool.InitStatic(gasTip, head, reserve)
+	pool.InitStatic(gasTip, head, nil, reserve)
 
 	// Start the reorg loop early, so it can handle requests generated during
 	// journal loading.
@@ -349,7 +349,7 @@ func (pool *LegacyPool) Init(gasTip uint64, head *types.Header, reserve txpool.A
 // InitStatic sets the gas price needed to keep a transaction in the pool
 // and the chain head to allow balance / nonce checks. This method doesn't
 // start loops or load journals, so the pool can be released automatically.
-func (pool *LegacyPool) InitStatic(gasTip uint64, head *types.Header, reserve txpool.AddressReserver) error {
+func (pool *LegacyPool) InitStatic(gasTip uint64, head *types.Header, statedb *state.StateDB, reserve txpool.AddressReserver) error {
 	// Set the address reserver to request exclusive access to pooled accounts
 	pool.reserve = reserve
 
@@ -359,12 +359,16 @@ func (pool *LegacyPool) InitStatic(gasTip uint64, head *types.Header, reserve tx
 	// Initialize the state with head block, or fallback to empty one in
 	// case the head state is not available (might occur when node is not
 	// fully synced).
-	statedb, err := pool.chain.StateAt(head.Root)
-	if err != nil {
-		statedb, err = pool.chain.StateAt(types.EmptyRootHash)
-	}
-	if err != nil {
-		return err
+	// If the state is provided, then use it directly.
+	var err error
+	if statedb == nil {
+		statedb, err = pool.chain.StateAt(head.Root)
+		if err != nil {
+			statedb, err = pool.chain.StateAt(types.EmptyRootHash)
+		}
+		if err != nil {
+			return err
+		}
 	}
 	pool.currentHead.Store(head)
 	pool.currentState = statedb

